haystack/schema.py

```diff
@@ -556,30 +576,261 @@ def append(self, key: str, value: pd.DataFrame):
         else:    
             self.node_results[key] = value
 
-    def calculate_metrics(self) -> Dict[str, float]:
+    def calculate_metrics(
+        self, 
+        simulated_top_k_reader: int = -1,
+        simulated_top_k_retriever: int = -1,
+        doc_relevance_col: str = "gold_id_match"
+    ) -> Dict[str, Dict[str, float]]:
         """
-            First dummy implementation of metrics calcuation just to show the way it's done.
-            TODO: implement retriever and reader specific metrics that must not rely on node names.
+        Calculates proper metrics for each node.
+        
+        Lower top_k values for reader and retriever than the actual values during the eval run can be simulated.
+        E.g. top_1_f1 for reader nodes can be calculated by setting simulated_top_k_reader=1.
+
+        Results for reader nodes with applied simulated_top_k_retriever should be considered with caution 
+        as there are situations the result can heavily differ from an actual eval run with corresponding top_k_retriever.
+
+        :param simulated_top_k_reader: simulates top_k param of reader
+        :param simulated_top_k_retriever: simulates top_k param of retriever.
+            remarks: there might be a discrepancy between simulated reader metrics and an actual pipeline run with retriever top_k
+        :param doc_relevance_col: column in the underlying eval table that contains the relevance criteria for documents.
+            values can be: 'gold_id_match', 'answer_match', 'gold_id_or_answer_match'
         """
-        reader_df = self.node_results["Reader"]
-        first_answers = reader_df[reader_df["rank"] == 1]
-        first_correct_answers = first_answers[first_answers.apply(
-            lambda x: x["answer"] in x["gold_answers"], axis=1)]
-
-        return {
-            "MatchInTop1": len(first_correct_answers) / len(first_answers) if len(first_answers) > 0 else 0.0
-        }
+        return {node: self._calculate_node_metrics(df, 
+                    simulated_top_k_reader=simulated_top_k_reader, 
+                    simulated_top_k_retriever=simulated_top_k_retriever,
+                    doc_relevance_col=doc_relevance_col) 
+            for node, df in self.node_results.items()}
+
+    def worst_queries(
```

Comment:
I'd suggest we rename this one to "wrong_examples" and later extend this method to support three different modes: 
1) the current one sorting by "metric"
2) random wrong examples (would be helpful to just iterate quickly over a few negatives to understand model behavior)
3) sorting by confidence (pred is wrong but model is highly confident) 

Later we could also add the equivalent "good_examples"

Reply:
I agree. @tstadel and I also talked about changing this method to show both bad and good examples. It's just switching between descending or ascending order when sorting the examples.

Reply:
renaming done